src/mono/mono/mini/method-to-ir.c

```diff
@@ -9546,6 +9558,63 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 				}
 			}
 
+			// ASSUME
+			//   interface ISomeIface { void Method(); }
+			//   struct SomeStruct : ISomeIface {...}
+			// OPTIMIZE
+			//   box SomeStruct
+			//   callvirt instance ISomeIface::Method()
+			// TO
+			//   call SomeStruct::Method()
+			guchar* callvirt_ip;
+			guint32 callvirt_proc_token;
+			if((callvirt_ip = il_read_callvirt(next_ip, end, &callvirt_proc_token)) && 
+				ip_in_bb(cfg, cfg->cbb, callvirt_ip) ) 
+			{
+				MonoMethod* iface_method = mini_get_method(cfg, method, callvirt_proc_token, NULL, generic_context);
+				MonoMethodSignature* iface_method_sig = mono_method_signature_internal(iface_method);
+
+				if(	val &&
+					val->flags != MONO_INST_FAULT && // not null
+					iface_method_sig != NULL && // callee signture is healthy
+					iface_method_sig->hasthis && 
+					iface_method_sig->param_count == 0) // the callee has no args (other than this)
```

Comment:
Do we assume 0 additional arguments for the sake of simplicity, or because it's not ok to optimize the call if there are additional arguments?

Reply:
Oh, it's because it won't actually be `box+callvirt` there will be other arguments in the way

Reply:
`this` is pushed on the stack first, so if there are any other arguments, there will be opcodes present that load them onto the stack between the `box` and `callvirt`. It would be difficult to match them - the JIT would have to look an unknown number of opcodes ahead.